Common - ArithTraits: moving from Kokkos to KokkosKernels

lucbv opened the pull request:
This simply moves the class to a new namespace to avoid having it in the Kokkos namespace.

lucbv committed:
Common - ArithTraits: moving from Kokkos to KokkosKernels

This simply moves the class to a new namespace to avoid having it
in the Kokkos namespace.

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

dalg24 commented on common/src/Kokkos_ArithTraits.hpp:
```diff
@@ -1644,9 +1644,15 @@ struct [[deprecated]] ArithTraits<qd_real> {
 namespace Details {
 template <typename T>
-using ArithTraits [[deprecated("Use Kokkos::ArithTraits instead")]] = ::Kokkos::ArithTraits<T>;
+using ArithTraits [[deprecated("Use KokkosKernels::ArithTraits instead")]] = ::KokkosKernels::ArithTraits<T>;
```
How long ago did you deprecate this alias?
Is removing it an option?

lucbv replied:
We are pretty close to 5.0.0, why would we do a backward breaking change against our policies at this point?

dalg24 replied:
I am talking about the alias in Details:: that was already marked as deprecated

lucbv replied:
I understand and what I said above still stands, do we really need to be removed before 5.0.0? That's only a few months away...

dalg24 replied:
What do you mean "before 5.0"?

lucbv replied:
I don't want to remove this until after the last release in the 4.7 series

dalg24 replied:
4.7 was the last minor release in the 4.x series

lucbv replied:
Okay so not 4.7.2 right? straight to 5.0.0?

dalg24 replied:
4.7.2 would come from 4.7.1 not develop

dalg24 reviewed (commented):
Any chance to deprecate the header as well and redirect to a new header with appropriate prefix?

lucbv committed:
Applying change of namespace to library and tests

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv committed:
Trying to fix the issue with sycl...

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv committed:
More fixes for std::complex<T>

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

dalg24 commented on common/src/Kokkos_ArithTraits.hpp:
```diff
@@ -1093,14 +1093,12 @@ class ArithTraits<std::complex<RealFloatType>> {
-    KOKKOS_IF_ON_HOST((using std::isinf;))
-    KOKKOS_IF_ON_DEVICE((using Kokkos::isinf;))
-    return isinf(real(x)) || isinf(imag(x));
+    return std::isinf(std::real(x)) || std::isinf(std::imag(x));
```
Why std::?

lucbv replied:
Honestly I am not sure I understand how the hole thing works on device but if you look further down, `real` and `imag` function are implemented by call `std::real` and `std::imag`.
So for now it's not worse than it was and it allows to complete the scope I defined for this PR.
Of course these functions could easily cause problem down the line and opening an issue to work on it is appropriate.

dalg24 commented:
> Any chance to deprecate the header as well and redirect to a new header with appropriate prefix?

Would you be willing to move the traits definition to a new `<KokkosKernels_ArithTraits.hpp>`header  and have `< Kokkos_ArithTraits.hpp >` just include it along with a using directive in the Kokkos:: namespace.

lucbv committed:
common: moving ArithTraits to KokkosKernels_ArithTraits.hpp

This avoid having us use the naming convention of Kokkos Core
for something that lives in Kokkos Kernels.
Next, I will modify header inclusions in the library to reflect
this change.

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv committed:
Updating header inclusion to use new KokkosKernels_ArithTraits.hpp

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

lucbv committed:
Applying clang-format

Signed-off-by: Luc Berger-Vergiat <lberge@sandia.gov>

ndellingwood commented:
We'll need to fix Trilinos for these changes, e.g.
```
In file included from /root/Trilinos/packages/stokhos/src/sacado/kokkos/vector/tpetra/Stokhos_Tpetra_MP_Vector.hpp:29:
/root/Trilinos/packages/stokhos/src/sacado/kokkos/vector/linalg/Kokkos_ArithTraits_MP_Vector.hpp:24:7: error: cannot specialize a dependent template
   24 | class ArithTraits< Sacado::MP::Vector<S> > {
```

ndellingwood commented:
@lucbv just to check, these changes only apply to versions >= 5.0 ? (So I know how to guard in Trilinos)

dalg24 commented:
I suppose you could do
```C++
#if __has_include(KokkosKernels_ArithTraits.hpp)  // since 5.0

lucbv commented:
Yes this would only apply to versions of Kokkos >= 5.0 so for a 4.7.2 release we would not want this change.
